File: 0875-koko-eating-bananas/0875-koko-eating-bananas.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        int low=1;
        int high=Integer.MIN_VALUE;
        for(int i=0;i<piles.length;i++){
            high=Math.max(high,piles[i]);
        }
        
        int ans= -1;
        while(low<=high){
            int mid=low+(high-low)/2;
            if(blackbox(mid,piles,h)){
                ans = mid;
                high=mid-1;
            }
            else
                low=mid+1;
        }
        return low;
    }
    public boolean blackbox(int maxpiles,int[] piles,int h){
        long hours=0;
        for(int i:piles){
            int time=i/maxpiles;
            hours+=time;
            if(i%maxpiles!=0) hours++;
        }
        if(hours<=h)
            return true;
        return false;
    }
}

File: 0875-koko-eating-bananas/0875-koko-eating-bananas.cpp
 class Solution {
     public:
    int minEatingSpeed(vector<int>& piles, int h) {
        int l = 1,m =0;
        for(int i=0;i<piles.size();i++){
            m = max(m,piles[i]);
        }
        while(l<=m){
            int mid = (l+m)/2;
            unsigned int tot = 0;
            cout<<mid<<" ";
            for(int i=0;i<piles.size();i++){
                
                tot+=(piles[i]/mid);
                
                if(piles[i]%mid)
                    tot++;
            }
            
            cout<<tot<<endl;
            if(tot>h){
                l = mid+1;
            }
            else
                m = mid-1;
        }
        return l;
    }

};